cheroot/workers/monitor.py

```diff
@@ -0,0 +1,154 @@
+            try:
+                self.function(*self.args, **self.kwargs)
+            except Exception as e:
+                self.log(
+                    'Error in background task thread function {}:{}.'.format(
+                        self.function,
+                        e,
+                    ),
+                )
+                # Quit on first error to avoid massive logs.
+                raise
+
+
+class Monitor:
+    """Monitor a BackgroundTask."""
+
+    callback = None
+    """The function to call at intervals."""
+
+    frequency = 60
+    """The time in seconds between callback runs."""
+
+    thread = None
+    """A :class:`BackgroundTask` thread."""
+    def __init__(self, callback, frequency=60, name=None, logger=None):
+        """Initialize Monitor."""
+        self.callback = callback
+        self.frequency = frequency
+        self.thread = None
+        self.name = name or self.__class__.__name__
+        self.log = logger or (lambda msg: None)
+
+    def start(self):
+        """Start our callback in its own background thread."""
+        if self.frequency <= 0:
+            self.log('Monitor thread recieved invalid frequency {}.'.format(
+                self.frequency,
+            ))
+            return
+        if self.thread is not None:
+            self.log('Monitor thread {} already started.'.format(self.name))
+            return
+        self.thread = BackgroundTask(
+            self.frequency,
+            self.callback,
+            logger=self.log,
+        )
+        self.thread.setName(self.name)
+        self.thread.start()
+        self.log('Started monitor thread {}.'.format(self.name))
+
+    def stop(self):
+        """Stop our callback's background task thread."""
+        if self.thread is None:
+            self.log('No thread running for {}.'.format(self.name))
+            return
+        if self.thread is not threading.currentThread():
+            name = self.thread.getName()
+            self.thread.cancel()
+            if not self.thread.daemon:
+                self.log('Joining {}'.format(name))
+                self.thread.join()
+            self.log('Stopped thread {}.'.format(name))
+        self.thread = None
+
+    def graceful(self):
+        """Stop the callback's background task thread and restart it."""
+        self.stop()
+        self.start()
+
+
+class ThreadPoolMonitor(Monitor):
+    """ThreadPoolMonitor for dynamic resizing."""
+
+    def __init__(self, frequency, name=None, logger=None):
+        """Initialize ThreadPoolMonitor."""
+        self._run = lambda: None
+        self._resizer = None
+        super(ThreadPoolMonitor, self).__init__(
+            self.run,
+            frequency=frequency,
+            name=name,
+            logger=logger,
+        )
+
+    def run(self):
+        """Run the monitor."""
+        self._run()
+
+    def configure(self, thread_pool, minspare, maxspare, shrinkfreq, logfreq):
+        """Configure the pool resizer."""
```

Comment:
It'd be useful to explain args with Sphinx syntax following this example: https://github.com/cherrypy/cheroot/blob/85a190f/cheroot/server.py#L1413-L1418